Docs/BazelForiOSDevelopers.md

```diff
@@ -0,0 +1,522 @@
+flags directly to the compiler. Please find canonical documentation on `copts`
+on the [Bazel
+docs](https://docs.bazel.build/versions/master/be/objective-c.html#objc_library.copts).
+Using `objc_library` to define compiler flags is useful for the per-rule level
+and many projects use macros and other layers of abstraction to [unify library
+level configuration](#Macros).
+
+
+## Extending Bazel
+
+### Starlark
+
+Bazel provides the
+[_pythonic_](https://stackoverflow.com/questions/25011078/what-does-pythonic-mean)
+programming language, Starlark. Starlark is used to implement build system
+logic and establish norms.  Generally, starlark calls into functionality that's
+implemented within Bazel. The coming segments [rules](#rules),
+[macros](#macros), and [aspects](#aspects) cover several examples.
+
+```
+.bzl file -> starlark -> bazel functions
+```
+
+Starlark enables extensibility without needing to fork the core build system.
+This boundary helps the developer to focus on their business logic instead of
+the core of the build system. For example, the Apple specific logic of bundling
+an `ios_application` is external to Bazel's core residing in the rule set,
+`rules_apple`.
+
+With Starlark, the possibilities are endless. _technically they aren't endless
+as the language is not turing complete_.For more information, please see the
+[Starlark
+documentation](https://docs.bazel.build/versions/master/skylark/language.html).
+
+
+### Rules
+
+Rules implement business logic for how the iOS application is built by creating
+actions. Actions represent invocations of external command line programs like
+`clang` or `bash`. Instances of rule are targets. Typical projects contain many
+rules, targets, and BUILD files.
+
+> A rule defines a series of actions that Bazel performs on inputs to produce a set of outputs.
+
+_- [the bazel documentation](https://docs.bazel.build/versions/master/skylark/rules.html)_
+
+
+In the previous segment, we created an iOS application with via a BUILD file.
+The `ios_application` rule is implemented by `rules_apple` and the
+`objc_library` is a native rule. Native rules are implemented in Java and have
+more power and capabilities needed to implement core functionality like remote
+execution and c++ compilation and are generally used as-is.
+
+```
+BUILD file -> target -> rule -> action -> execution
+```
+
+Bazel and open source rules should provide most functionality to build an iOS
+application. Generally, defining custom rules isn't required but can improve and
+consolidate functionality. To learn more about how to create custom rules, the
+document, [Bazel
+extensions](https://docs.bazel.build/versions/master/skylark/concepts.html)
+contains a comprehensive overview.
+
+### Macros
+
+Like [rules](#Rules) and [aspects](#Aspects), macros are
+defined in `.bzl` files.  A macro is a convenient way to call a rule, and not
+recognized by Bazel in the same way a rule is.
```

Comment:
Can you elaborate on "not recognized by Bazel in the same way"?  Maybe with an example

Reply:
@jszumski thanks for the suggestion, this is a really good point. Here's my quick explanation. 

> As Bazel builds targets, declared by rules, it can't "build" a macro. A macro is evaluated during the evaluation of a `BUILD` file and instantiates rules. Once the macro is translated into a rule, the macro is irrelevant to the build process. In C++ compilation, this behavior is similar to the pre-processor feature of the compiler. Please see the following example of wrapping `objc_library` 

Bazel also has a definition of this which we can link https://docs.bazel.build/versions/master/skylark/concepts.html#macros-and-rules